# Exploring eBay Car Sales Data

In this project, we'll work with a dataset of used cars from *eBay Kleinanzeigen*, a classifieds section of the German eBay website. The original dataset isn't available anymore, but you can find it [here](https://data.world/data-society/used-cars-data).

The aim of this project is to clean the data and analize the included user cars listings. Let's start by importing the the libraries we need and reading the dataset into pandas.

In [1]:
import numpy as np
import pandas as pd

Read the `autos.csv`  CSV file into pandas, and assign it to the variable name `autos`.
    
* Try without specifying any encoding (which will default to `UTF-8`)
* If you get an encoding error, try the next two most popular encodings (`Latin-1` and `Windows-1252`) until you are able to read the file without error.

In [2]:
try: 
    autos = pd.read_csv("autos.csv")
except UnicodeDecodeError:
    print("Error")

Error


In [3]:
try: 
    autos = pd.read_csv("autos.csv",encoding = "Latin-1")
except UnicodeDecodeError:
    print("Error")

In [4]:
try: 
    autos = pd.read_csv("autos.csv",encoding = "Windows-1252")
except UnicodeDecodeError:
    print("Error")

In [5]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

Use the `DataFrame.info()`and `DataFrame.head()` methods to print information about the autos dataframe, as well the first rows.

In [6]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [7]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

From the previous work, we can make the following observations:

* The dataset contains 20 columns, most of wich are strings.
* Some columns have null values, but none have more than ~20% null values.
* The column names use *camelcase* instead of Python's preferred *snakecase*.
* The data is in german language

Let's convert the column names from *camelcase* to *snakecase* and reword some of the column name based on the data sictionary to be more descriptive.

In [8]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [9]:
autos.rename({'dateCrawled':"date_crawled",
               'offerType':'offer_type',
             "vehicleType":"vehicle_type",
             "yearOfRegistration":"registration_year",
             "powerPS":"power_ps",
             'monthOfRegistration':"registration_month",
             'fuelType':'fuel_type',
             'notRepairedDamage':"unrepaired_damage",
             'dateCreated':"ad_created",
             'nrOfPictures':'nr_of_pictures',
             'postalCode':'postal_code',
             'lastSeen':'last_seen'}, axis=1, inplace = True)

In [10]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

* Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
* Examples of numeric data stored as text which can be cleaned and converted.

`DataFrame.describe(include = "all")` to get both numeric and categorical columns

In [11]:
autos.describe(include = "all")

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

In [12]:
autos.describe()

price  registration_year       power_ps      kilometer  \
count  3.715280e+05      371528.000000  371528.000000  371528.000000   
mean   1.729514e+04        2004.577997     115.549477  125618.688228   
std    3.587954e+06          92.866598     192.139578   40112.337051   
min    0.000000e+00        1000.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      70.000000  125000.000000   
50%    2.950000e+03        2003.000000     105.000000  150000.000000   
75%    7.200000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        9999.000000   20000.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count       371528.000000        371528.0  371528.00000  
mean             5.734445             0.0   50820.66764  
std              3.712412             0.0   25799.08247  
min              0.000000             0.0    1067.00000  
25%              3.000000             0.0   30459.00000  
50%              6.000000             0.0   49610.00000  
75%              9.000000             0.0   71546.00000  
max             12.000000             0.0   99998.00000

In [13]:
autos["seller"].value_counts()

privat        371525
gewerblich         3
Name: seller, dtype: int64

In [14]:
autos["offer_type"].value_counts()

Angebot    371516
Gesuch         12
Name: offer_type, dtype: int64

In [15]:
autos["abtest"].value_counts()

test       192585
control    178943
Name: abtest, dtype: int64

In [16]:
autos["vehicle_type"].value_counts()

limousine     95894
kleinwagen    80023
kombi         67564
bus           30201
cabrio        22898
coupe         19015
suv           14707
andere         3357
Name: vehicle_type, dtype: int64

In [17]:
autos["gearbox"].value_counts()

manuell      274214
automatik     77105
Name: gearbox, dtype: int64

In [18]:
print(f"The {round(77105/autos.shape[0]*100,2)}% of the cars have an automatic gearbox")

The 20.75% of the cars have an automatic gearbox


In [19]:
autos["fuel_type"].value_counts()

benzin     223857
diesel     107746
lpg          5378
cng           571
hybrid        278
andere        208
elektro       104
Name: fuel_type, dtype: int64

In [20]:
autos["unrepaired_damage"].value_counts()

nein    263182
ja       36286
Name: unrepaired_damage, dtype: int64

In [21]:
print(f"The {round(36286/autos.shape[0]*100,2)}% of the cars have unrepaired damage")

The 9.77% of the cars have unrepaired damage


In [22]:
autos["date_crawled"] = pd.to_datetime(autos["date_crawled"])
autos["ad_created"] = pd.to_datetime(autos["ad_created"])
autos["last_seen"] = pd.to_datetime(autos["last_seen"])

In [23]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   date_crawled        371528 non-null  datetime64[ns]
 1   name                371528 non-null  object        
 2   seller              371528 non-null  object        
 3   offer_type          371528 non-null  object        
 4   price               371528 non-null  int64         
 5   abtest              371528 non-null  object        
 6   vehicle_type        333659 non-null  object        
 7   registration_year   371528 non-null  int64         
 8   gearbox             351319 non-null  object        
 9   power_ps            371528 non-null  int64         
 10  model               351044 non-null  object        
 11  kilometer           371528 non-null  int64         
 12  registration_month  371528 non-null  int64         
 13  fuel_type           338142 no

We learned that there are a number of text columns where almost all of the values are the same (`seller` and `offer_type`). Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `kilometer` and `price` columns. Here's the steps we'll take:
* Analize the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

In [24]:
autos["kilometer"].unique().shape

(13,)

In [25]:
autos["price"].unique().shape

(5597,)

In [26]:
autos[["kilometer","price"]].describe()

kilometer         price
count  371528.000000  3.715280e+05
mean   125618.688228  1.729514e+04
std     40112.337051  3.587954e+06
min      5000.000000  0.000000e+00
25%    125000.000000  1.150000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  2.147484e+09

In [27]:
autos["kilometer"].value_counts().sort_index(ascending = True)

5000        7069
10000       1949
20000       5676
30000       6041
40000       6376
50000       7615
60000       8669
70000       9773
80000      11053
90000      12523
100000     15920
125000     38067
150000    240797
Name: kilometer, dtype: int64

In [28]:
autos["price"].value_counts().sort_index(ascending = False).head(5)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
Name: price, dtype: int64

In [29]:
autos["price"].value_counts().sort_index(ascending = True).head()

0    10778
1     1189
2       12
3        8
4        1
Name: price, dtype: int64

In [30]:
autos_copy = autos.copy()

In [31]:
autos_copy["price"].value_counts(normalize=True).sort_index(ascending = True).loc[1000:25000].sum()

0.7483904308692749

In [32]:
autos_copy = autos_copy[autos_copy["price"].between(1000,25000)]

In [33]:
autos_copy["price"].shape

(278048,)

In [34]:
autos_copy.drop(["seller","offer_type"],axis = 1,inplace = True)

In [35]:
autos_copy

date_crawled                                   name  price  \
1      2016-03-24 10:58:45                   A5_Sportback_2.7_Tdi  18300   
2      2016-03-14 12:52:21         Jeep_Grand_Cherokee_"Overland"   9800   
3      2016-03-17 16:54:04                     GOLF_4_1_4__3TÜRER   1500   
4      2016-03-31 17:25:20         Skoda_Fabia_1.4_TDI_PD_Classic   3600   
6      2016-04-01 20:48:51            Peugeot_206_CC_110_Platinum   2200   
...                    ...                                    ...    ...   
371521 2016-03-27 20:36:20     Opel_Zafira_1.6_Elegance_TÜV_12/16   1150   
371523 2016-03-14 17:48:27             Suche_t4___vito_ab_6_sitze   2200   
371524 2016-03-05 19:56:21  Smart_smart_leistungssteigerung_100ps   1199   
371525 2016-03-19 18:57:12     Volkswagen_Multivan_T4_TDI_7DC_UY2   9200   
371526 2016-03-20 19:41:08                 VW_Golf_Kombi_1_9l_TDI   3400   

         abtest vehicle_type  registration_year    gearbox  power_ps  \
1          test        coupe               2011    manuell       190   
2          test          suv               2004  automatik       163   
3          test   kleinwagen               2001    manuell        75   
4          test   kleinwagen               2008    manuell        69   
6          test       cabrio               2004    manuell       109   
...         ...          ...                ...        ...       ...   
371521  control          bus               2000    manuell         0   
371523     test          NaN               2005        NaN         0   
371524     test       cabrio               2000  automatik       101   
371525     test          bus               1996    manuell       102   
371526     test        kombi               2002    manuell       100   

              model  kilometer  registration_month fuel_type           brand  \
1               NaN     125000                   5    diesel            audi   
2             grand     125000                   8    diesel            jeep   
3              golf     150000                   6    benzin      volkswagen   
4             fabia      90000                   7    diesel           skoda   
6           2_reihe     150000                   8    benzin         peugeot   
...             ...        ...                 ...       ...             ...   
371521       zafira     150000                   3    benzin            opel   
371523          NaN      20000                   1       NaN  sonstige_autos   
371524       fortwo     125000                   3    benzin           smart   
371525  transporter     150000                   3    diesel      volkswagen   
371526         golf     150000                   6    diesel      volkswagen   

       unrepaired_damage ad_created  nr_of_pictures  postal_code  \
1                     ja 2016-03-24               0        66954   
2                    NaN 2016-03-14               0        90480   
3                   nein 2016-03-17               0        91074   
4                   nein 2016-03-31               0        60437   
6                   nein 2016-04-01               0        67112   
...                  ...        ...             ...          ...   
371521              nein 2016-03-27               0        26624   
371523               NaN 2016-03-14               0        39576   
371524              nein 2016-03-05               0        26135   
371525              nein 2016-03-19               0        87439   
371526               NaN 2016-03-20               0        40764   

                 last_seen  
1      2016-04-07 01:46:50  
2      2016-04-05 12:47:46  
3      2016-03-17 17:40:17  
4      2016-04-06 10:17:21  
6      2016-04-05 18:18:39  
...                    ...  
371521 2016-03-29 10:17:23  
371523 2016-04-06 00:46:52  
371524 2016-03-11 18:17:12  
371525 2016-04-07 07:15:26  
371526 2016-03-24 12:45:21  

[278048 rows x 18 columns]

Let's now move on to the date columns and understand the date range the data covers.

In [36]:
autos_copy[["date_crawled","ad_created","last_seen"]].value_counts(normalize=True,dropna = False).sort_index()

date_crawled         ad_created  last_seen          
2016-03-05 14:06:23  2016-03-05  2016-03-06 18:29:35    0.000004
2016-03-05 14:06:24  2016-03-05  2016-03-24 09:16:44    0.000004
                                 2016-04-06 22:15:33    0.000004
2016-03-05 14:06:25  2016-03-05  2016-03-09 15:19:01    0.000004
                                 2016-04-06 22:15:40    0.000004
                                                          ...   
2016-04-07 14:36:53  2016-04-07  2016-04-07 14:36:53    0.000004
2016-04-07 14:36:55  2016-04-07  2016-04-07 14:36:55    0.000004
2016-04-07 14:36:56  2016-04-07  2016-04-07 14:36:56    0.000004
2016-04-07 14:36:57  2016-04-07  2016-04-07 14:36:57    0.000004
2016-04-07 14:36:58  2016-04-07  2016-04-07 14:36:58    0.000004
Length: 272925, dtype: float64

In [37]:
autos_copy["registration_year"].describe()

count    278048.000000
mean       2004.880776
std          59.006270
min        1000.000000
25%        2001.000000
50%        2005.000000
75%        2009.000000
max        9999.000000
Name: registration_year, dtype: float64

Let's count the number of listings with cars that fall outside the 1900-2016 interval and see if it's saffe to remove those rows entirely, or if we need more custom logic.

In [38]:
round(autos_copy["registration_year"].value_counts(normalize=True).sort_index().loc[1900:2016].sum()*100,2)

96.15

From the previous analysis we'll keep with the values between 1951 and 2016 years. Some of the reason because this interval was choose are:
* A car can't be first registered after the listing was seen, any vehicle with a registration year
* This data represent the 96.15 % of the total.

Explore the unique values in `brand` column, and decide on which brands you want to agggregate by.
* You might want to select the top 20, or you might want to select those that have over a certain percentage of the total values (e.g. > 5%).

Create an empty dictionary to hold your aggregate data.
* Loop over your selected brands, and assign the mean price to the dictionary, with the brand name as the key.

To identify the unique brands, we can use the `Series.unique()` method. This method returns an array of unique values from any series, Then, we can loop over that array and perform our operation.

In [39]:
len(autos_copy["brand"].unique())

40

Selecting the top 20 brands.

In [40]:
autos_copy["brand"].value_counts().head(20)

volkswagen        60379
bmw               33659
mercedes_benz     29904
audi              26654
opel              25425
ford              16523
renault           10874
peugeot            8252
fiat               6110
seat               5172
skoda              5167
smart              4751
toyota             4185
mazda              4156
citroen            3943
nissan             3435
mini               3281
hyundai            3093
volvo              2420
sonstige_autos     2413
Name: brand, dtype: int64

Selecting those brands that have over a certain percentage of the total. (e.g. 5%)

In [41]:
brands = autos_copy["brand"].value_counts(normalize=True).head(6)*100
brands

volkswagen       21.715315
bmw              12.105464
mercedes_benz    10.754978
audi              9.586115
opel              9.144105
ford              5.942499
Name: brand, dtype: float64

In [42]:
brand_mean_price = autos_copy[autos_copy["brand"].isin(["volkswagen","bmw","mercedes_benz","audi","opel","ford"])]\
                                .groupby("brand").price\
                                .mean()\
                                .sort_values(ascending = False)
brand_mean_price_d = dict(brand_mean_price)
brand_mean_price_d

{'audi': 7978.176746454566,
 'bmw': 7646.803440387415,
 'mercedes_benz': 7295.040262172284,
 'volkswagen': 5931.013663691018,
 'ford': 4828.722144888942,
 'opel': 4129.160668633235}

We aggregated across brands to understand mean price. We observed that in the top 6 brands, there's a distinct price gap.
* Audi, BMW, and Mercedes Benz are more expensive.
* Ford and Opel are less expensive.
* Volkswagen is in between.

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:
* it's difficult to compare more than two aggregated series objects if we want to extend to more columns.
* we can't compare more than a few rows from each series object.


In [43]:
brand_mean_mileage = autos_copy[autos_copy["brand"].isin(["volkswagen","bmw","mercedes_benz","audi","opel","ford"])]\
                                .groupby("brand").kilometer\
                                .mean()\
                                .sort_values(ascending = False)
brand_mean_mileage_d = dict(brand_mean_mileage)
brand_mean_mileage_d

{'bmw': 135861.4337918536,
 'audi': 133617.6558865461,
 'mercedes_benz': 133468.59951845906,
 'volkswagen': 127324.7321088458,
 'opel': 123473.94296951819,
 'ford': 120616.4134842341}

In [44]:
bmp_series = pd.Series(brand_mean_price_d)
bmp_df = pd.DataFrame(bmp_series,columns = ["mean_price"])
bmp_df

mean_price
audi           7978.176746
bmw            7646.803440
mercedes_benz  7295.040262
volkswagen     5931.013664
ford           4828.722145
opel           4129.160669

In [45]:
bmp_series = pd.Series(brand_mean_mileage_d)
bmp_series

bmw              135861.433792
audi             133617.655887
mercedes_benz    133468.599518
volkswagen       127324.732109
opel             123473.942970
ford             120616.413484
dtype: float64

In [46]:
bmp_df["average_mileage"] = bmp_series
bmp_df

mean_price  average_mileage
audi           7978.176746    133617.655887
bmw            7646.803440    135861.433792
mercedes_benz  7295.040262    133468.599518
volkswagen     5931.013664    127324.732109
ford           4828.722145    120616.413484
opel           4129.160669    123473.942970

In this guided project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings. Here are some stepsfor you to consider:
* Data cleaning next steps:
    
    - Identify categorical data that uses german words, translate them and map the values to their english counterparts.
    - Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
    - See if there are a particular keywords in the name column that you can extract as new columns.

* Analysis next steps:

    - Find the most common brand/model combination.
    - Split the kilometer column into groups, and use aggregation to see if averages prices follows any patterns based on the mileage.
    - How much cheaper are cars with damage than their non-damaged counterparts?
    

In [47]:
autos_copy.head()

date_crawled                            name  price abtest  \
1 2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  18300   test   
2 2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"   9800   test   
3 2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER   1500   test   
4 2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic   3600   test   
6 2016-04-01 20:48:51     Peugeot_206_CC_110_Platinum   2200   test   

  vehicle_type  registration_year    gearbox  power_ps    model  kilometer  \
1        coupe               2011    manuell       190      NaN     125000   
2          suv               2004  automatik       163    grand     125000   
3   kleinwagen               2001    manuell        75     golf     150000   
4   kleinwagen               2008    manuell        69    fabia      90000   
6       cabrio               2004    manuell       109  2_reihe     150000   

   registration_month fuel_type       brand unrepaired_damage ad_created  \
1                   5    diesel        audi                ja 2016-03-24   
2                   8    diesel        jeep               NaN 2016-03-14   
3                   6    benzin  volkswagen              nein 2016-03-17   
4                   7    diesel       skoda              nein 2016-03-31   
6                   8    benzin     peugeot              nein 2016-04-01   

   nr_of_pictures  postal_code           last_seen  
1               0        66954 2016-04-07 01:46:50  
2               0        90480 2016-04-05 12:47:46  
3               0        91074 2016-03-17 17:40:17  
4               0        60437 2016-04-06 10:17:21  
6               0        67112 2016-04-05 18:18:39

In [48]:
autos_copy["date_crawled"] = autos_copy["date_crawled"].dt.strftime("%Y%m%d")
autos_copy["ad_created"] = autos_copy["ad_created"].dt.strftime("%Y%m%d")
autos_copy["last_seen"] = autos_copy["last_seen"].dt.strftime("%Y%m%d")

In [49]:
autos_copy["gearbox"] = autos_copy["gearbox"].map({'manuell':"manual","automatik":"automatic"})
autos_copy["unrepaired_damage"] = autos_copy["unrepaired_damage"].map({'ja':"yes","nein":"no"})
autos_copy["fuel_type"] = autos_copy["fuel_type"].map({'benzin':"petrol","andere":"other","elektro":"electric"})

In [50]:
autos_copy.drop(["nr_of_pictures"],axis=1,inplace = True)

In [51]:
autos_copy.describe(include = "all")

date_crawled      name          price  abtest vehicle_type  \
count        278048    278048  278048.000000  278048       257588   
unique           34    170573            NaN       2            8   
top        20160403  BMW_318i            NaN    test    limousine   
freq          11060       561            NaN  144074        76204   
mean            NaN       NaN    6049.020759     NaN          NaN   
std             NaN       NaN    5209.862056     NaN          NaN   
min             NaN       NaN    1000.000000     NaN          NaN   
25%             NaN       NaN    2150.000000     NaN          NaN   
50%             NaN       NaN    4100.000000     NaN          NaN   
75%             NaN       NaN    8299.000000     NaN          NaN   
max             NaN       NaN   25000.000000     NaN          NaN   

        registration_year gearbox       power_ps   model      kilometer  \
count       278048.000000  267751  278048.000000  266433  278048.000000   
unique                NaN       2            NaN     250            NaN   
top                   NaN  manual            NaN    golf            NaN   
freq                  NaN  204513            NaN   22662            NaN   
mean          2004.880776     NaN     123.365833     NaN  124771.676113   
std             59.006270     NaN     193.886073     NaN   39135.259183   
min           1000.000000     NaN       0.000000     NaN    5000.000000   
25%           2001.000000     NaN      78.000000     NaN  100000.000000   
50%           2005.000000     NaN     116.000000     NaN  150000.000000   
75%           2009.000000     NaN     150.000000     NaN  150000.000000   
max           9999.000000     NaN   20000.000000     NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage ad_created  \
count        278048.000000    159120      278048            235838     278048   
unique                 NaN         3          40                 2        108   
top                    NaN    petrol  volkswagen                no   20160403   
freq                   NaN    158945       60379            218792      11130   
mean              6.020615       NaN         NaN               NaN        NaN   
std               3.599402       NaN         NaN               NaN        NaN   
min               0.000000       NaN         NaN               NaN        NaN   
25%               3.000000       NaN         NaN               NaN        NaN   
50%               6.000000       NaN         NaN               NaN        NaN   
75%               9.000000       NaN         NaN               NaN        NaN   
max              12.000000       NaN         NaN               NaN        NaN   

          postal_code last_seen  
count   278048.000000    278048  
unique            NaN        34  
top               NaN  20160406  
freq              NaN     63780  
mean     51425.524607       NaN  
std      25688.811740       NaN  
min       1067.000000       NaN  
25%      31135.000000       NaN  
50%      50389.000000       NaN  
75%      72202.000000       NaN  
max      99998.000000       NaN

In [52]:
autos_copy["unrepaired_damage"].fillna("no",inplace = True)
autos_copy["fuel_type"].fillna("petrol",inplace = True)

In [53]:
autos_copy["model"].value_counts()

golf          22662
andere        18962
3er           17127
a4             8635
passat         8475
              ...  
charade           3
samara            3
serie_3           3
serie_1           1
elefantino        1
Name: model, Length: 250, dtype: int64

In [54]:
import re
pattern = r"_(?P<model_name>\w+)_(.+)_(.+)?"
name_parts = autos_copy["name"].str.extract(pattern)
name_parts["model_name"].str.lower()

1               sportback
2                   grand
3                     4_1
4                   fabia
6                  206_cc
               ...       
371521             zafira
371523       t4___vito_ab
371524              smart
371525    multivan_t4_tdi
371526       golf_kombi_1
Name: model_name, Length: 278048, dtype: object

In [55]:
autos_copy["model"].fillna(name_parts["model_name"].str.lower(),inplace=True)

In [56]:
autos_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 278048 entries, 1 to 371526
Data columns (total 17 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   date_crawled        278048 non-null  object
 1   name                278048 non-null  object
 2   price               278048 non-null  int64 
 3   abtest              278048 non-null  object
 4   vehicle_type        257588 non-null  object
 5   registration_year   278048 non-null  int64 
 6   gearbox             267751 non-null  object
 7   power_ps            278048 non-null  int64 
 8   model               274808 non-null  object
 9   kilometer           278048 non-null  int64 
 10  registration_month  278048 non-null  int64 
 11  fuel_type           278048 non-null  object
 12  brand               278048 non-null  object
 13  unrepaired_damage   278048 non-null  object
 14  ad_created          278048 non-null  object
 15  postal_code         278048 non-null  int64 
 16  la

In [57]:
autos_copy["gearbox"].value_counts()

manual       204513
automatic     63238
Name: gearbox, dtype: int64

In [58]:
autos_copy[["brand","gearbox"]].value_counts(normalize = True,dropna = False).sort_index().head(20)

brand       gearbox  
alfa_romeo  automatic    0.000403
            manual       0.005596
            NaN          0.000212
audi        automatic    0.031009
            manual       0.061680
            NaN          0.003172
bmw         automatic    0.040615
            manual       0.077235
            NaN          0.003204
chevrolet   automatic    0.001974
            manual       0.003607
            NaN          0.000227
chrysler    automatic    0.002136
            manual       0.001568
            NaN          0.000101
citroen     automatic    0.001859
            manual       0.011710
            NaN          0.000611
dacia       automatic    0.000004
            manual       0.003068
dtype: float64

In [59]:
autos_copy["gearbox"].fillna(method='bfill',inplace = True)

In [60]:
autos_copy["gearbox"].value_counts()

manual       212345
automatic     65703
Name: gearbox, dtype: int64

In [61]:
autos_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 278048 entries, 1 to 371526
Data columns (total 17 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   date_crawled        278048 non-null  object
 1   name                278048 non-null  object
 2   price               278048 non-null  int64 
 3   abtest              278048 non-null  object
 4   vehicle_type        257588 non-null  object
 5   registration_year   278048 non-null  int64 
 6   gearbox             278048 non-null  object
 7   power_ps            278048 non-null  int64 
 8   model               274808 non-null  object
 9   kilometer           278048 non-null  int64 
 10  registration_month  278048 non-null  int64 
 11  fuel_type           278048 non-null  object
 12  brand               278048 non-null  object
 13  unrepaired_damage   278048 non-null  object
 14  ad_created          278048 non-null  object
 15  postal_code         278048 non-null  int64 
 16  la

In [62]:
#autos_copy.dropna(inplace = True)

In [63]:
popular_cars = autos_copy["name"].value_counts().head(20)

In [64]:
pck = popular_cars.keys()

In [65]:
autos_no = autos_copy[(autos_copy["name"].isin(pck)) & (autos_copy["unrepaired_damage"] =="no")]

In [66]:
autos_no_n=autos_no[["name","unrepaired_damage","price"]].groupby("name").mean()
autos_no_n

price
name                                          
Audi_A4_Avant_1.9_TDI              3797.076555
Audi_A4_Avant_2.0_TDI_DPF          8996.880531
BMW_116i                           7951.945055
BMW_316i                           2461.419540
BMW_318i                           4313.079545
BMW_320i                           4994.089686
Ford_Fiesta                        2861.552995
MINI_Mini_Cooper                   7172.425439
MINI_Mini_One                      6903.481308
Opel_Corsa                         2899.198020
Opel_Corsa_1.0_12V                 2198.581395
Opel_Corsa_1.2_16V                 2891.246154
Peugeot_206                        1850.713568
Volkswagen_Golf                    5050.165079
Volkswagen_Golf_1.4                2477.885650
Volkswagen_Golf_1.6                2522.652921
Volkswagen_Golf_1.9_TDI            3373.987603
Volkswagen_Passat_Variant_1.9_TDI  2957.982609
Volkswagen_Polo                    2469.477352
Volkswagen_Polo_1.2                3223.834239

In [67]:
autos_yes = autos_copy[(autos_copy["name"].isin(pck)) & (autos_copy["unrepaired_damage"] =="yes")]
autos_yes_y=autos_yes[["name","unrepaired_damage","price"]].groupby("name").mean()
autos_yes_y

price
name                                          
Audi_A4_Avant_1.9_TDI              2672.000000
Audi_A4_Avant_2.0_TDI_DPF          5860.703704
BMW_116i                           5305.793103
BMW_316i                           1889.333333
BMW_318i                           2409.575758
BMW_320i                           2882.416667
Ford_Fiesta                        2128.888889
MINI_Mini_Cooper                   4158.125000
MINI_Mini_One                      4511.947368
Opel_Corsa                         2304.454545
Opel_Corsa_1.0_12V                 1687.210526
Opel_Corsa_1.2_16V                 2158.434783
Peugeot_206                        1599.894737
Volkswagen_Golf                    2841.906250
Volkswagen_Golf_1.4                2260.606061
Volkswagen_Golf_1.6                1944.160000
Volkswagen_Golf_1.9_TDI            2921.750000
Volkswagen_Passat_Variant_1.9_TDI  2117.541667
Volkswagen_Polo                    1714.285714
Volkswagen_Polo_1.2                2488.257143

In [68]:
autos_no_n["unrepaired_damage_yes"] = autos_yes_y

In [72]:
autos_no_n.rename({"price":"unrepaired_damage_no"},axis=1,inplace=True)

In [76]:
autos_no_n.drop("yes",axis=1,inplace=True)

In [78]:
autos_no_n["diff"] = autos_no_n["unrepaired_damage_no"] - autos_no_n["unrepaired_damage_yes"]

In [82]:
autos_no_n["diff_perc"] = round(100*(autos_no_n["diff"]/autos_no_n["unrepaired_damage_no"]),2) 

In [83]:
autos_no_n

unrepaired_damage_no  \
name                                                      
Audi_A4_Avant_1.9_TDI                       3797.076555   
Audi_A4_Avant_2.0_TDI_DPF                   8996.880531   
BMW_116i                                    7951.945055   
BMW_316i                                    2461.419540   
BMW_318i                                    4313.079545   
BMW_320i                                    4994.089686   
Ford_Fiesta                                 2861.552995   
MINI_Mini_Cooper                            7172.425439   
MINI_Mini_One                               6903.481308   
Opel_Corsa                                  2899.198020   
Opel_Corsa_1.0_12V                          2198.581395   
Opel_Corsa_1.2_16V                          2891.246154   
Peugeot_206                                 1850.713568   
Volkswagen_Golf                             5050.165079   
Volkswagen_Golf_1.4                         2477.885650   
Volkswagen_Golf_1.6                         2522.652921   
Volkswagen_Golf_1.9_TDI                     3373.987603   
Volkswagen_Passat_Variant_1.9_TDI           2957.982609   
Volkswagen_Polo                             2469.477352   
Volkswagen_Polo_1.2                         3223.834239   

                                   unrepaired_damage_yes         diff  \
name                                                                    
Audi_A4_Avant_1.9_TDI                        2672.000000  1125.076555   
Audi_A4_Avant_2.0_TDI_DPF                    5860.703704  3136.176827   
BMW_116i                                     5305.793103  2646.151951   
BMW_316i                                     1889.333333   572.086207   
BMW_318i                                     2409.575758  1903.503788   
BMW_320i                                     2882.416667  2111.673019   
Ford_Fiesta                                  2128.888889   732.664107   
MINI_Mini_Cooper                             4158.125000  3014.300439   
MINI_Mini_One                                4511.947368  2391.533940   
Opel_Corsa                                   2304.454545   594.743474   
Opel_Corsa_1.0_12V                           1687.210526   511.370869   
Opel_Corsa_1.2_16V                           2158.434783   732.811371   
Peugeot_206                                  1599.894737   250.818831   
Volkswagen_Golf                              2841.906250  2208.258829   
Volkswagen_Golf_1.4                          2260.606061   217.279590   
Volkswagen_Golf_1.6                          1944.160000   578.492921   
Volkswagen_Golf_1.9_TDI                      2921.750000   452.237603   
Volkswagen_Passat_Variant_1.9_TDI            2117.541667   840.440942   
Volkswagen_Polo                              1714.285714   755.191638   
Volkswagen_Polo_1.2                          2488.257143   735.577096   

                                   diff_perc  
name                                          
Audi_A4_Avant_1.9_TDI                  29.63  
Audi_A4_Avant_2.0_TDI_DPF              34.86  
BMW_116i                               33.28  
BMW_316i                               23.24  
BMW_318i                               44.13  
BMW_320i                               42.28  
Ford_Fiesta                            25.60  
MINI_Mini_Cooper                       42.03  
MINI_Mini_One                          34.64  
Opel_Corsa                             20.51  
Opel_Corsa_1.0_12V                     23.26  
Opel_Corsa_1.2_16V                     25.35  
Peugeot_206                            13.55  
Volkswagen_Golf                        43.73  
Volkswagen_Golf_1.4                     8.77  
Volkswagen_Golf_1.6                    22.93  
Volkswagen_Golf_1.9_TDI                13.40  
Volkswagen_Passat_Variant_1.9_TDI      28.41  
Volkswagen_Polo                        30.58  
Volkswagen_Polo_1.2                    22.82